cogs/permissions.py

```diff
@@ -0,0 +1,439 @@
+                        row = await cursor.fetchone()
+                        if row:
+                            permission_config = PermissionConfig.get_config(
+                                is_enabled=row[2], whitelisted_channel_ids=frozenset(loads(row[3])))
+                        else:
+                            permission_config = PermissionConfig.get_default_config_for_permission(permission)
+                self.cache[server_id][permission.id] = permission_config
+
+            return self.cache[server_id][permission.id]
+
+    async def save_permission_config_for_server(
+            self, server_id: int, permission: Permission, permission_config: PermissionConfig):
+        if permission_config == PermissionConfig.get_default_config_for_permission(permission):
+            # Setting to the default config is effectively the same as resetting the config, so favor the simpler path.
+            await self.reset_permission_config_for_server(server_id, permission)
+            return
+
+        async with self.cache_lock:
+            if server_id not in self.cache:
+                self.cache[server_id] = {}
+
+            self.cache[server_id][permission.id] = permission_config
+
+            async with connect(self.db) as connection:
+                is_enabled = permission_config.is_enabled
+                whitelisted_channel_ids = dumps(sorted(permission_config.whitelisted_channel_ids))
+                query = 'SELECT * FROM permissions WHERE server_id=? AND permission_id=?'
+                async with connection.execute(query, (server_id, permission.id)) as cursor:
+                    if await cursor.fetchone():
+                        await cursor.execute('UPDATE permissions '
+                                             'SET is_enabled=?,'
+                                             '    whitelisted_channel_ids=? '
+                                             'WHERE server_id=? AND permission_id=?',
+                                             (is_enabled, whitelisted_channel_ids, server_id, permission.id))
+                    else:
+                        await cursor.execute('INSERT INTO permissions VALUES (?, ?, ?, ?)',
+                                             (server_id, permission.id, is_enabled, whitelisted_channel_ids))
+                await connection.commit()
+
+    async def reset_permission_config_for_server(self, server_id: int, permission: Permission):
+        async with self.cache_lock:
+            if server_id not in self.cache:
+                self.cache[server_id] = {}
+
+            self.cache[server_id][permission.id] = PermissionConfig.get_default_config_for_permission(permission)
+
+            async with connect(self.db) as connection:
+                from_statement = 'FROM permissions WHERE server_id=? AND permission_id=?'
+                async with connection.execute(f'SELECT * {from_statement}', (server_id, permission.id)) as cursor:
```

Comment:
I see no reason for this check, sqlite will just return the number of affected rows (e.g. 0 or 1) and will not throw any warnings/errors if the row doesnt exist